File: exemplesPAP/src/tp00/MyMath.java
package tp00;

import java.util.List;

/**
 * 
 * @author Didier
 *
 */
public class MyMath {

	/**
	 * Calcule la factorielle de n
	 * 
	 * @param n entier positif
	 * @return n!
	 */
	public static long factorielle(int n) {
		assert n >= 0 && n < 21 : "n doit être un nombre positif";

		long y = 1;
		for (int i = 2; i <= n; i++)
			y = y * i;
		return y;
	}

	/**
	 * Calcule la factorielle de n
	 * 
	 * @param n entier positif <21
	 * @return n!
	 */
	public static long factorielle2(int n) {
		assert n >= 0 && n < 21 : "Le nombre doit être positif et inférieur à 21";
		long res = 1;
		// n est déjà initialisé
		for (; n > 1; n--)
			res *= n;
		return res;
	}

	/**
	 * Calcule x^y
	 * 
	 * @param x réel
	 * @param y entier positif
	 * @return
	 */
	public static double power(double x, int y) {
		assert y >= 0 : "Y doit être un entier positif";
		// Elimine les cas exposant 0
		if (y == 0)
			return 1.0;
		// Y>0
		double res = x;
		for (int i = 2; i <= y; i++)
			res = res * x;

		return res;
	}

	/**
	 * Calcule x^y
	 * 
	 * @param x réel
	 * @param y entier positif
	 * @return
	 */
	public static double power2(double x, int y) {
		assert y >= 0 : "Y doit être un entier positif";
		// Elimine les cas exposant 0
		if (y == 0)
			return 1.0;
		if (y == 1)
			return x;
		// Y>1
		double res = x;
		int z = y >> 1;// z=y/2
		for (int i = 2; i <= z; i++)
			res *= x;
		res *= res;
		// Si le nombre est impaire, je mutiplie par x
		if ((y & 1) == 1) // ou (y % 2 !=0)
			res *= x;
		return res;
	}

	/**
	 * 
	 * @param a un entier >0
	 * @param b un entier >0
	 * @return le pgcd de a et b
	 */
	public static int pgcd(int a, int b) {
		assert a > 0 && b > 0 : "a et b doivent être >0";
		int reste;
		while (b != 0) {
			reste = a % b;
			a = b;
			b = reste;
		}
		return a;
	}

	/**
	 * Calcule le nombre de bits à 1
	 * 
	 * @param n un entier
	 * @return nb de bits à 1
	 */
	public static int nbUn(long n) {
		int cpt = 0;
		while (n != 0) {
			if ((n & 1) > 0)
				cpt++;
			n = n >>> 1;
		}
		return cpt;
	}

	/**
	 * Calcule le nombre de bytes pour coder ce caractère en UTF8
	 * 
	 * @param c le caractère
	 * @return nombre de bytes
	 */
	public static int nbBytesUtf8(char c) {
		// Type casting pour avoir le code du caractère
		int res;
		int n = (int) c;
		if ((n & 0xFF80) == 0)
			res = 1;
		else if ((n & 0xF800) == 0)
			res = 2;
		else
			res = 3;
		return res;
	}

	/**
	 * Vérifie si un nombre est premier
	 * 
	 * @param n entier >1
	 * @return true si premier
	 */
	public static boolean estPremier(int n) {
		assert n > 1 : "n doit être un entier positif >1";
		boolean premier = n == 2 || n % 2 != 0;
		int d = 3;
		final int FIN = (int) Math.sqrt(n);
		while (premier && d <= FIN) {
			premier = n % d != 0;
			d = d + 2;
		}
		return premier;
	}

	public static void main(String[] args) {
		System.out.println(factorielle(20));
		System.out.println(factorielle2(3));
		System.out.println("---------------------------------");

		List<Integer> listePremier = List.of(1, 2, 3, 5, 7, 11, 13, 17);
		long sum = 1;
		for (Integer i : listePremier) {
			sum += i;
			System.out.println("somme= " + sum);
		}

		if (listePremier != null && listePremier.get(0) < 2)
			System.out.println("bic presque vide");
	}

}

File: exemplesPAP/python/mymath.py

def factorielle(n: int)-> int:
    ''' Calcule la factorielle de n  '''
    assert type(n)== int and n>=0, "n doit être un entier positif"
    y=1
    for n in range(n,1,-1):
        y=y*n
    return y # Retourne n!

def power(x: float,y: int = 2)->float:
    '''Calcule x^y avec y entier positif '''
    assert  isinstance(x, (int, float))  and isinstance(y, int) and y>= 0 , 'Erreur x: réel, y:entier >=0'
    
    if y==0 :
        return 1
    # Y>0
    res = x
    for i in range(1,y):
        res = res * x
        
    return res

def power2(x: float,y: int)-> float:
    '''Calcule x^y avec y entier'''
    assert isinstance(x, (int, float))  and isinstance(y, int), 'Erreur x: réel, y:entier'
    
    if y==0 : return 1
    # Y>0
    yPositif= y>0
    y= abs(y)
    if y==1 or x==0 or x==1: return x if yPositif else 1/x
    # Y est positif maintenant
    res = x
    for i in range(1,y//2):
        res = res * x
    res*= res
    #si y est impaire alors res<-res*x
    if y%2 != 0:
        res*=x
        
    return res if yPositif else 1/res

def pgcd(a: int,b: int)-> int:
    '''Calcule le pgcd des nombres a>0 et b>0 '''
    assert type(a)== int and type(b)== int and a>0 and b>0, " a et b doivent être > 0"
    while b!=0:
        a, b = b, a%b
    return a

def nbUnV1(n: int)-> int:
    '''Calcule le nombre de bits à 1'''
    assert type(n) == int and n>=0, 'Le nombre doit être un entier positif'
    cpt = 0
    masque = 1
    taille = n.bit_length()
    for i in range(taille):
        if (n & masque) > 0 :
            cpt = cpt + 1
        masque = masque <<1
    return cpt, taille

def nbUn(n: int)-> int:
    '''Calcule le nombre de bits à 1'''
    assert type(n) == int and n>=0, 'Le nombre doit être un entier positif'
    cpt = 0
    while n != 0:
        if (n & 1) > 0 :
            cpt = cpt + 1
        n = n >> 1
    return cpt

def nbBytesUtf8(c : str)-> int:
    '''Retourne le nombre de bytes pour coder ce car en utf8'''
    assert type(c) == str and len(c) == 1, "Doit être un seul caractère"
    n = ord(c) #n <-- code unicode du caractère
    if n & 0xFF80 == 0 :
        res = 1
    elif n & 0xF800 == 0 :
        res = 2
    else: res = 3
    
    return res

def nbBytesUtf8_V2(c : str)-> int:
    '''Retourne le nombre de bytes pour coder ce car en utf8'''
    assert type(c) == str and len(c) == 1, "Doit être un seul caractère"
    n = ord(c) #n <-- code unicode du caractère
    if n & 0xFF80 == 0 :
        return  1
    if n & 0xF800 == 0 :
        return  2
    return  3

#Exemples utilisant un nombre variable d'aguments
def affiche(*chaine,separateur=','):
    '''Exemple d'appel avec un nombre variable
       d'arguments ==> liste'''
    for c in chaine:
        print(c,end=separateur)
    print()

def connectionDb(**param):
    '''Exemple d'appel avec un nombre variable
       d'arguments nommés ==> dictionnaire'''
    for e in list(param.items()):
       print(f'clé: {e[0]} ==> valeur: {e[1]}')

#Variable globale au module      
cpt = 0

def cube(a):
    global cpt 
    cpt = cpt + 1 #Incrément le cpt (indique le nbr d'appels)
    return a * a *a

#A exécuter uniquement lorsqu'il s'agit du module main
if __name__ == "__main__":
    
    connectionDb(login='admin',password='adminPW')
    affiche(1,2,'T',4,5,'_')
    affiche(1,2,'T',4,5,separateur='_')
    car='€'
    print(f'Le caractère {car} à besoin de {nbBytesUtf8(car)} bytes')
    print(f'Nombre de bits à 1 de 6 est {nbUn(6)} ')
    power2(2,-1)
    d=int(input("Entrez un nombre: "))
    print(d,"  ",type(d))
    res=factorielle(d)
    print(f' La factorielle de {d} est {res} ')
    print(f' le pgcd de 15 et 9 = {pgcd(15,9)}')
    
    